File: app-cdk/temp/python/60-observability/cicd_workshop/app-cdk/app_cdk/pipeline_cdk_stack.py
import os
from constructs import Construct
from aws_cdk import (
    Stack,
    CfnOutput,
    aws_codecommit as codecommit,
    aws_codepipeline as codepipeline,
    aws_codebuild as codebuild,
    aws_codepipeline_actions as codepipeline_actions,
    aws_iam as iam,
    aws_codedeploy as codedeploy,
    aws_cloudwatch as cloudwatch,
    Duration as duration,
    aws_sns as sns,
    aws_sns_subscriptions as subscriptions,
    aws_events as events,
    aws_events_targets as targets
)


class PipelineCdkStack(Stack):

    def __init__(self, scope: Construct, id: str, ecr_repository, test_app_fargate, prod_app_fargate, green_target_group, green_load_balancer_listener, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        # Pipeline code will go here
        repo = codecommit.Repository(
            self, 'CICD_Workshop',
            repository_name='CICD_Workshop',
            description='Repository for my application code and infrastructure'
        )

        pipeline = codepipeline.Pipeline(
            self, 'CICD_Pipeline',
            cross_account_keys=False
        )

        code_quality_build = codebuild.PipelineProject(
            self, 'Code Quality',
            build_spec=codebuild.BuildSpec.from_source_filename('./buildspec_test.yml'),
            environment=codebuild.BuildEnvironment(
                build_image=codebuild.LinuxBuildImage.STANDARD_5_0,
                privileged=True,
                compute_type=codebuild.ComputeType.LARGE,
            ),
        )

        docker_build_project = codebuild.PipelineProject(
            self, 'Docker Build',
            build_spec=codebuild.BuildSpec.from_source_filename('./buildspec_docker.yml'),
            environment=codebuild.BuildEnvironment(
                build_image=codebuild.LinuxBuildImage.STANDARD_5_0,
                privileged=True,
                compute_type=codebuild.ComputeType.LARGE,
                environment_variables={
                    'IMAGE_TAG': codebuild.BuildEnvironmentVariable(
                        type=codebuild.BuildEnvironmentVariableType.PLAINTEXT,
                        value='latest'
                    ),
                    'IMAGE_REPO_URI': codebuild.BuildEnvironmentVariable(
                        type=codebuild.BuildEnvironmentVariableType.PLAINTEXT,
                        value=ecr_repository.repository_uri
                    ),
                    'AWS_DEFAULT_REGION': codebuild.BuildEnvironmentVariable(
                        type=codebuild.BuildEnvironmentVariableType.PLAINTEXT,
                        value=os.environ['CDK_DEFAULT_REGION']
                    )
                }
            ),
        )

        docker_build_project.add_to_role_policy(iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=[
                'ecr:GetAuthorizationToken',
                'ecr:BatchCheckLayerAvailability',
                'ecr:GetDownloadUrlForLayer',
                'ecr:GetRepositoryPolicy',
                'ecr:DescribeRepositories',
                'ecr:ListImages',
                'ecr:DescribeImages',
                'ecr:BatchGetImage',
                'ecr:InitiateLayerUpload',
                'ecr:UploadLayerPart',
                'ecr:CompleteLayerUpload',
                'ecr:PutImage'
            ],
            resources=['*'],
        ))

        source_output = codepipeline.Artifact()
        unit_test_output = codepipeline.Artifact()
        docker_build_output = codepipeline.Artifact()

        source_action = codepipeline_actions.CodeCommitSourceAction(
            action_name='CodeCommit',
            repository=repo,
            output=source_output,
            branch='main'
        )

        pipeline.add_stage(
            stage_name='Source',
            actions=[source_action]
        )

        build_action = codepipeline_actions.CodeBuildAction(
            action_name='Unit-Test',
            project=code_quality_build,
            input=source_output,
            outputs=[unit_test_output]
        )

        pipeline.add_stage(
            stage_name='Code-Quality-Testing',
            actions=[build_action]
        )

        docker_build_action = codepipeline_actions.CodeBuildAction(
            action_name='Docker-Build',
            project=docker_build_project,
            input=source_output,
            outputs=[docker_build_output]
        )

        pipeline.add_stage(
            stage_name='Docker-Build',
            actions=[docker_build_action]
        )

        pipeline.add_stage(
            stage_name='Deploy-Test',
            actions=[
                codepipeline_actions.EcsDeployAction(
                    action_name='Deploy-Test',
                    service=test_app_fargate.service,
                    input=docker_build_output
                )
            ]
        )

        ecs_code_deploy_app = codedeploy.EcsApplication(self, "my-app",
            application_name="my-app"
        )

        prod_ecs_deployment_group = codedeploy.EcsDeploymentGroup(self, "my-app-dg",
            service=prod_app_fargate.service,
            blue_green_deployment_config=codedeploy.EcsBlueGreenDeploymentConfig(
                blue_target_group=prod_app_fargate.target_group,
                green_target_group=green_target_group,
                listener=prod_app_fargate.listener,
                test_listener=green_load_balancer_listener
            ),
            deployment_config=codedeploy.EcsDeploymentConfig.LINEAR_10_PERCENT_EVERY_1_MINUTES,
            application=ecs_code_deploy_app
        )

        pipeline.add_stage(
            stage_name='Deploy-Production',
            actions=[
                codepipeline_actions.ManualApprovalAction(
                action_name='Approve-Prod-Deploy',
                run_order=1
              ),
              codepipeline_actions.CodeDeployEcsDeployAction(
                  action_name='ABlueGreen-deployECS',
                  deployment_group=prod_ecs_deployment_group,
                  app_spec_template_input=source_output,
                  task_definition_template_input=source_output,
                  run_order=2
              )
            ]
        )
        CfnOutput(
            self, 'CodeCommitRepositoryUrl',
            value=repo.repository_clone_url_grc
        )

        build_rate = cloudwatch.GraphWidget(
        title="Build Successes and Failures",
        width=6,
        height=6,
        view=cloudwatch.GraphWidgetView.PIE,
        left=[
            cloudwatch.Metric(
                namespace="AWS/CodeBuild",
                metric_name="SucceededBuilds",
                statistic='sum',
                label='Succeeded Builds',
                period=duration.days(30)
            ),
            cloudwatch.Metric(
                namespace="AWS/CodeBuild",
                metric_name="FailedBuilds",
                statistic='sum',
                label='Failed Builds',
                period=duration.days(30)
            )
        ]
        )

        builds_count = cloudwatch.SingleValueWidget(
            title="Successes and Failures",
            width=6,
            height=6,
            metrics=[
                cloudwatch.Metric(
                    namespace="AWS/CodeBuild",
                    metric_name="Builds",
                    statistic='sum',
                    label='Builds',
                    period=duration.days(30)
                )
            ]
        )

        average_duration = cloudwatch.GaugeWidget(
            title="Average Build Time",
            width=6,
            height=6,
            metrics=[
                cloudwatch.Metric(
                    namespace="AWS/CodeBuild",
                    metric_name="Duration",
                    statistic='Average',
                    label='Duration',
                    period=duration.hours(1)
                )
            ],
            left_y_axis={
                'min': 0,
                'max': 300,
            }
        )

        queued_duration = cloudwatch.GaugeWidget(
            title="Build Queue Duration",
            width=6,
            height=6,
            metrics=[
                cloudwatch.Metric(
                    namespace="AWS/CodeBuild",
                    metric_name="QueuedDuration",
                    statistic='Average',
                    label='Duration',
                    period=duration.hours(1)
                )
            ],
            left_y_axis={
                'min': 0,
                'max': 60,
            }
        )

        download_duration = cloudwatch.GraphWidget(
        title="Checkout Duration",
        width=24,
        height=5,
        left=[
            cloudwatch.Metric(
                namespace="AWS/CodeBuild",
                metric_name="DownloadSourceDuration",
                statistic='max',
                label='Duration',
                period=duration.minutes(5),
                color=cloudwatch.Color.PURPLE
            )
        ]
        )
  
        dashboard = cloudwatch.Dashboard(
            self, 'CICD_Dashboard',
            dashboard_name='CICD_Dashboard',
            widgets=[
                [build_rate, builds_count, average_duration, queued_duration, download_duration]
            ]
        )
        # Create SNS topic for failures
        failure_topic = sns.Topic(
            self, "BuildFailure",
            display_name="BuildFailure"
        )

        # Create a subscription for email
        email_subscription = subscriptions.EmailSubscription('YOUR_EMAIL_ADDRESS@your_provider.com')

        # Add the email subscription to the failureTopic
        failure_topic.add_subscription(email_subscription)

        # CloudWatch Event Rule for pipeline failures
        pipeline_failure_rule = events.Rule(self, "PipelineFailureRule",
          description="Notify on pipeline failures",
          event_pattern={
            "source": ["aws.codepipeline"],
            "detail_type": ["CodePipeline Pipeline Execution State Change"],
            "detail": {
                "state": ["FAILED"]
            }
          }
        )

        # Add SNS topic as a target with input transformer
        pipeline_failure_rule.add_target(targets.SnsTopic(
          failure_topic,
          message=events.RuleTargetInput.from_text(
            f"Pipeline Failure Detected! Pipeline: {events.EventField.from_path('$.detail.pipeline')} "
            f"Execution ID: {events.EventField.from_path('$.detail.execution-id')}"
          )
        ))

File: app-cdk/temp/typescript/60-observability/cicd_workshop/app-cdk/lib/pipeline-cdk-stack.ts
import { Stack, StackProps, CfnOutput, Duration } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as codecommit from 'aws-cdk-lib/aws-codecommit';
import * as codepipeline from 'aws-cdk-lib/aws-codepipeline';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as codepipeline_actions from 'aws-cdk-lib/aws-codepipeline-actions';
import * as ecr from 'aws-cdk-lib/aws-ecr';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as ecsPatterns from 'aws-cdk-lib/aws-ecs-patterns';
import * as codedeploy from 'aws-cdk-lib/aws-codedeploy';
import * as elbv2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import { Rule, EventPattern } from 'aws-cdk-lib/aws-events';
import * as events from 'aws-cdk-lib/aws-events';


interface ConsumerProps extends StackProps {
  ecrRepository: ecr.Repository,
  fargateServiceTest: ecsPatterns.ApplicationLoadBalancedFargateService,
  greenTargetGroup: elbv2.ApplicationTargetGroup,
  greenLoadBalancerListener: elbv2.ApplicationListener,
  fargateServiceProd: ecsPatterns.ApplicationLoadBalancedFargateService,
}

export class PipelineCdkStack extends Stack {
  constructor(scope: Construct, id: string, props: ConsumerProps) {
    super(scope, id, props);

    const sourceRepo = new codecommit.Repository(this, 'CICD_Workshop', {
      repositoryName: 'CICD_Workshop',
      description: 'Repository for my application code and infrastructure',
    });

    const pipeline = new codepipeline.Pipeline(this, 'Pipeline', {
      pipelineName: 'CICD_Pipeline',
      crossAccountKeys: false,
    });

    const codeBuild = new codebuild.PipelineProject(this, 'CodeBuild', {
      environment: {
        buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
        privileged: true,
        computeType: codebuild.ComputeType.LARGE,
      },
      buildSpec: codebuild.BuildSpec.fromSourceFilename('buildspec_test.yml'),
    });

    const dockerBuild = new codebuild.PipelineProject(this, 'DockerBuild', {
      environmentVariables: {
        IMAGE_TAG: { value: 'latest' },
        IMAGE_REPO_URI: { value: props.ecrRepository.repositoryUri },
        AWS_DEFAULT_REGION: { value: process.env.CDK_DEFAULT_REGION },
      },
      environment: {
        buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
        privileged: true,
        computeType: codebuild.ComputeType.LARGE,
      },
      buildSpec: codebuild.BuildSpec.fromSourceFilename('buildspec_docker.yml'),
    });

    const dockerBuildRolePolicy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      resources: ['*'],
      actions: [
        'ecr:GetAuthorizationToken',
        'ecr:BatchCheckLayerAvailability',
        'ecr:GetDownloadUrlForLayer',
        'ecr:GetRepositoryPolicy',
        'ecr:DescribeRepositories',
        'ecr:ListImages',
        'ecr:DescribeImages',
        'ecr:BatchGetImage',
        'ecr:InitiateLayerUpload',
        'ecr:UploadLayerPart',
        'ecr:CompleteLayerUpload',
        'ecr:PutImage',
      ],
    });

    dockerBuild.addToRolePolicy(dockerBuildRolePolicy);

    const sourceOutput = new codepipeline.Artifact();
    const unitTestOutput = new codepipeline.Artifact();
    const dockerBuildOutput = new codepipeline.Artifact();

    pipeline.addStage({
      stageName: 'Source',
      actions: [
        new codepipeline_actions.CodeCommitSourceAction({
          actionName: 'CodeCommit',
          repository: sourceRepo,
          output: sourceOutput,
          branch: 'main',
        }),
      ],
    });

    pipeline.addStage({
      stageName: 'Code-Quality-Testing',
      actions: [
        new codepipeline_actions.CodeBuildAction({
          actionName: 'Unit-Test',
          project: codeBuild,
          input: sourceOutput,
          outputs: [unitTestOutput],
        }),
      ],
    });

    pipeline.addStage({
      stageName: 'Docker-Push-ECR',
      actions: [
        new codepipeline_actions.CodeBuildAction({
          actionName: 'Docker-Build',
          project: dockerBuild,
          input: sourceOutput,
          outputs: [dockerBuildOutput],
        }),
      ],
    });

    pipeline.addStage({
      stageName: 'Deploy-Test',
      actions: [
        new codepipeline_actions.EcsDeployAction({
          actionName: 'Deploy-Fargate-Test',
          service: props.fargateServiceTest.service,
          input: dockerBuildOutput,
        }),
      ],
    });

    const ecsCodeDeployApp = new codedeploy.EcsApplication(this, "my-app", { applicationName: 'my-app' });
    const prodEcsDeploymentGroup = new codedeploy.EcsDeploymentGroup(this, "my-app-dg", {
      service: props.fargateServiceProd.service,
      blueGreenDeploymentConfig: {
        blueTargetGroup: props.fargateServiceProd.targetGroup,
        greenTargetGroup: props.greenTargetGroup,
        listener: props.fargateServiceProd.listener,
        testListener: props.greenLoadBalancerListener
      },
      deploymentConfig: codedeploy.EcsDeploymentConfig.LINEAR_10PERCENT_EVERY_1MINUTES,
      application: ecsCodeDeployApp,
    });
    pipeline.addStage({
      stageName: 'Deploy-Production',
      actions: [
        new codepipeline_actions.ManualApprovalAction({
          actionName: 'Approve-Prod-Deploy',
          runOrder: 1
        }),
        new codepipeline_actions.CodeDeployEcsDeployAction({
          actionName: 'BlueGreen-deployECS',
          deploymentGroup: prodEcsDeploymentGroup,
          appSpecTemplateInput: sourceOutput,
          taskDefinitionTemplateInput: sourceOutput,
          runOrder: 2
        })
      ]
    });
    new CfnOutput(this, 'CodeCommitRepositoryUrl', {
      value: sourceRepo.repositoryCloneUrlGrc,
    });
    const buildRate = new cloudwatch.GraphWidget({
      title: 'Build Successes and Failures',
      width: 6,
      height: 6,
      view: cloudwatch.GraphWidgetView.PIE,
      left: [
        new cloudwatch.Metric({
          namespace: 'AWS/CodeBuild',
          metricName: 'SucceededBuilds',
          statistic: 'sum',
          label: 'Succeeded Builds',
          period: Duration.days(30),
        }),
        new cloudwatch.Metric({
          namespace: 'AWS/CodeBuild',
          metricName: 'FailedBuilds',
          statistic: 'sum',
          label: 'Failed Builds',
          period: Duration.days(30),
        }),
      ],
    });
    const buildsCount = new cloudwatch.SingleValueWidget({
      title: 'Total Builds',
      width: 6,
      height: 6,
      metrics: [
        new cloudwatch.Metric({
          namespace: 'AWS/CodeBuild',
          metricName: 'Builds',
          statistic: 'sum',
          label: 'Builds',
          period: Duration.days(30),
        }),
      ],
    });
    const averageDuration = new cloudwatch.GaugeWidget({
      title: 'Average Build Time',
      width: 6,
      height: 6,
      metrics: [
        new cloudwatch.Metric({
          namespace: 'AWS/CodeBuild',
          metricName: 'Duration',
          statistic: 'avg',
          label: 'Duration',
          period: Duration.hours(1),
        }),
      ],
      leftYAxis: {
        min: 0,
        max: 300,
      },
    });
    const queuedDuration = new cloudwatch.GaugeWidget({
      title: 'Build Queue Duration',
      width: 6,
      height: 6,
      metrics: [
        new cloudwatch.Metric({
          namespace: 'AWS/CodeBuild',
          metricName: 'QueuedDuration',
          statistic: 'avg',
          label: 'Duration',
          period: Duration.hours(1),
        }),
      ],
      leftYAxis: {
        min: 0,
        max: 60,
      },
    });
    const downloadDuration = new cloudwatch.GraphWidget({
      title: 'Checkout Duration',
      width: 24,
      height: 5,
      left: [
        new cloudwatch.Metric({
          namespace: 'AWS/CodeBuild',
          metricName: 'DownloadSourceDuration',
          statistic: 'max',
          label: 'Duration',
          period: Duration.minutes(5),
          color: cloudwatch.Color.PURPLE,
        }),
      ],
    });
    new cloudwatch.Dashboard(this, 'CICD_Dashboard', {
      dashboardName: 'CICD_Dashboard',
      widgets: [
        [
          buildRate,
          buildsCount,
          averageDuration,
          queuedDuration,
          downloadDuration,
        ],
      ],
    });
    const failureTopic = new sns.Topic(this, "BuildFailure", {
      displayName: "BuildFailure",
    });
    const emailSubscription = new subscriptions.EmailSubscription('YOUR_EMAIL_ADDRESS@your_provider.com');
    failureTopic.addSubscription(emailSubscription);
    // CloudWatch event rule triggered on pipeline failures
    const pipelineFailureRule = new Rule(this, 'PipelineFailureRule', {
      description: 'Notify on pipeline failures',
      eventPattern: {
        source: ['aws.codepipeline'],
        detailType: ['CodePipeline Pipeline Execution State Change'],
        detail: {
          state: ['FAILED']
        }
      }
    });
    // Add SNS topic as a target
    pipelineFailureRule.addTarget(new targets.SnsTopic(failureTopic, {
      message: events.RuleTargetInput.fromText(`Pipeline Failure Detected! Pipeline: ${events.EventField.fromPath('$.detail.pipeline')}, Execution ID: ${events.EventField.fromPath('$.detail.execution-id')}`),
    }));
  }
}
